shared-module/audiofilters/Distortion.c

```diff
@@ -0,0 +1,374 @@
+            } else {
+                // For unsigned samples set to the middle which is "quiet"
+                if (MP_LIKELY(self->bits_per_sample == 16)) {
+                    memset(word_buffer, 32768, length * (self->bits_per_sample / 8));
+                } else {
+                    memset(hword_buffer, 128, length * (self->bits_per_sample / 8));
+                }
+            }
+
+            length = 0;
+        } else {
+            // we have a sample to play and apply effect
+            // Determine how many bytes we can process to our buffer, the less of the sample we have left and our buffer remaining
+            uint32_t n = MIN(self->sample_buffer_length, length);
+
+            int16_t *sample_src = (int16_t *)self->sample_remaining_buffer; // for 16-bit samples
+            int8_t *sample_hsrc = (int8_t *)self->sample_remaining_buffer; // for 8-bit samples
+
+            if (mix <= 0.01) { // if mix is zero pure sample only
+                for (uint32_t i = 0; i < n; i++) {
+                    if (MP_LIKELY(self->bits_per_sample == 16)) {
+                        word_buffer[i] = sample_src[i];
+                    } else {
+                        hword_buffer[i] = sample_hsrc[i];
+                    }
+                }
+            } else {
+
+                uint32_t word_mask = 0xFFFFFFFF ^ ((1 << (uint32_t)roundf(drive * 14.0)) - 1); // LOFI mode bit mask
+
+                for (uint32_t i = 0; i < n; i++) {
+                    int32_t sample_word = 0;
+                    if (MP_LIKELY(self->bits_per_sample == 16)) {
+                        sample_word = sample_src[i];
+                    } else {
+                        if (self->samples_signed) {
+                            sample_word = sample_hsrc[i];
+                        } else {
+                            // Be careful here changing from an 8 bit unsigned to signed into a 32-bit signed
+                            sample_word = (int8_t)(((uint8_t)sample_hsrc[i]) ^ 0x80);
+                        }
+                    }
+
+                    int32_t word = sample_word * pre_gain;
+                    switch (self->mode) {
+                        case DISTORTION_MODE_CLIP: {
+                            mp_float_t word_sign = word < 0 ? -1.0f : 1.0f;
+                            word = powf(fabs(word / 32768.0), 1.0001 - drive) * word_sign * 32767.0;
+                            word = MIN(MAX(word, -32767), 32768); // Hard clip
+                        } break;
+                        case DISTORTION_MODE_LOFI: {
+                            word = word & word_mask;
+                        } break;
+                        case DISTORTION_MODE_OVERDRIVE: {
+                            mp_float_t x = word / 32768.0 * 0.686306;
+                            mp_float_t z = 1 + expf(sqrtf(fabs(x)) * -0.75);
+                            word = (expf(x) - expf(-x * z)) / (expf(x) + expf(-x)) * 32767.0;
```

Comment:
does it actually perform well enough to call complicated floating-point math functions 7 times per sample? I am .. stunned, actually.

Reply:
Ha ha, yes it does on the RP2350. But of course, this is an area which needs work. I'd like to move these operations over to integer calculations where possible.

Reply:
I've optimized a few areas of this algorithm where it was feasible to reduce floating point calculations, but `audiofilters.DistortionMode.OVERDRIVE` requires the most by far. My simple testing so far has been promising, but if processing time is still a concern, I think we'll have to research an alternative algorithm to implement and move away from this one or even remove this mode altogether in favor of `audiofilters.DistortionMode.CLIP`.